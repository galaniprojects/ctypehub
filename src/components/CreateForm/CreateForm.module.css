.container {
  composes: bigContainer from '../Container.module.css';
}

.form {
  max-inline-size: 50rem;
  color: var(--color-text-secondary);
  font-size: 0.875rem;
}

.heading {
  font-size: 1.375rem;
  margin: 0;
}

.anchor {
  color: inherit;
}

.label {
  composes: label from '../Input/Input.module.css';
}

.input {
  composes: component from '../Input/Input.module.css';
}

.description {
  border: 1px solid var(--color-border);
  border-radius: var(--border-radius);
  box-sizing: border-box;
  inline-size: 100%;
  block-size: 6rem;
  font-family: inherit;
  font-size: inherit;
  padding: 0.5rem 0.75rem;
}

.fieldset {
  composes: fieldset from '../Input/Input.module.css';
}

.add {
  padding: 0.375rem 1rem;
  font-family: inherit;
  font-size: inherit;
  cursor: pointer;
  inline-size: 100%;
  block-size: 2.25rem;
  text-align: start;
  border-radius: var(--border-radius);
  background: buttonface url('./plus-solid.svg') no-repeat right 1rem
    center/0.75rem;
}

.tags {
  box-sizing: border-box;
  border-radius: var(--border-radius);
  border: 1px solid var(--color-border);
  font-family: inherit;
  padding: 0.25rem 0.75rem;
  background: white;
  margin: 0;
  list-style: none;
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  gap: 0.25rem;
}

.tag {
  box-sizing: border-box;
  display: inline-flex;
  align-items: center;
  background: var(--color-background-primary);
  color: white;
  border-radius: var(--border-radius);
  padding-inline-start: 0.5rem;
  padding-block: 0.25rem;
  gap: 0.25rem;
  overflow-wrap: anywhere;
}

.removeTag {
  border: 0;
  padding: 0;
  inline-size: 1.5rem;
  block-size: 1.125rem;
  align-self: flex-start;
  background: url('./xmark-solid.svg') no-repeat center/50%;
  display: inline-flex;
  cursor: pointer;
  flex-shrink: 0;
}

.tagInputContainer {
  flex-grow: 1;
  flex-shrink: 0;
}

.tagInput {
  border: none;
  inline-size: 100%;
  block-size: 1.5rem;
  font-size: inherit;
  font-family: inherit;
}

.tagInputDescription {
  margin: 0;
  font-size: 0.75rem;
}

.submit {
  composes: primary from '../Button.module.css';
}

.progress {
  color: var(--color-text-primary);
}

@keyframes rotation {
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(359deg);
  }
}

.spinner {
  --size: calc(74rem / 16);

  block-size: var(--size);
  inline-size: var(--size);
  display: block;
  margin-block-end: 0.625rem;
  margin-inline: auto;
  background: url('./spinner.svg') no-repeat center center;
  animation: rotation 1.5s infinite linear;
}

.output {
  display: block;
  margin-block-end: 1.25rem;
  padding-block-end: 7.5rem;
  background: url('./circle-xmark.svg') no-repeat center bottom/6.25rem;
  text-align: center;
  font-size: 1.375rem;
  color: var(--color-text-primary);
}

.retry {
  composes: primary from '../Button.module.css';
  inline-size: 100%;
}

@media screen and (width >= 600px) {
  .form {
    font-size: 1rem;
  }
}

@media screen and (width >= 1000px) {
  .container {
    background: var(--color-background-secondary) url('./big-logo.svg')
      no-repeat top 14rem right -3rem;
  }
}
