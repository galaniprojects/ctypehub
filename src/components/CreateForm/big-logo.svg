<svg xmlns="http://www.w3.org/2000/svg" width="616" height="616"><g fill="none" fill-rule="nonzero" opacity=".499"><path fill="#3E4C60" d="M272.274 427.81c25.188 7.132 51.518 6.09 75.452-2.117 6.257-2.148 13.125.85 15.843 6.86l65.287 144.118c3.008 6.652-.129 14.458-6.916 17.183-69.807 27.955-145.742 28.868-214.1 5.995-6.981-2.34-10.616-9.954-7.979-16.799l57.036-147.483c2.38-6.139 9.055-9.553 15.409-7.758h-.032ZM3.257 367.981c6.987 34.602 20.097 68.594 39.65 100.425 19.585 31.862 44.01 58.918 71.768 80.817 5.754 4.533 14.12 3.396 18.399-2.531l92.49-128.17c3.847-5.335 3.078-12.768-1.89-17.077-19.633-17.06-33.048-39.487-39.218-63.836-1.619-6.392-7.87-10.477-14.392-9.452l-156.23 24.59c-7.228 1.137-12.02 8.09-10.577 15.25v-.016ZM13.011 207.78A306.17 306.17 0 0 0 .138 261.454c-1.061 7.22 4.13 13.866 11.427 14.601l157.646 15.879c6.557.655 12.616-3.706 13.886-10.144a124.892 124.892 0 0 1 2.427-10.112c1.8-6.294-1.656-12.892-7.795-15.256L29.854 199.856c-6.846-2.62-14.497.99-16.843 7.907v.016ZM65.35 132.942l128.567 92.619c5.334 3.838 12.788 3.099 17.093-1.879 2.297-2.65 4.691-5.17 7.181-7.595 4.723-4.592 5.077-12.06.932-17.181L119.488 75.762c-4.61-5.7-13.044-6.375-18.539-1.526a306.914 306.914 0 0 0-38.17 40.288c-4.546 5.749-3.39 14.13 2.555 18.418h.016ZM185.126 36.292l64.713 144.124c2.685 5.992 9.478 9.036 15.68 6.905a121.12 121.12 0 0 1 9.988-2.98c6.345-1.618 10.357-7.898 9.334-14.37L260.244 13.83c-1.135-7.226-8.055-12.016-15.216-10.575a304.155 304.155 0 0 0-52.934 15.845c-6.777 2.724-9.957 10.526-6.968 17.19ZM335.912 11.53l-15.847 157.185c-.657 6.57 3.765 12.578 10.239 13.844a125.555 125.555 0 0 1 35.765 12.867c11.506 6.233 21.52 13.892 30.141 22.736 4.615 4.727 12.067 5.128 17.194.978l122.84-99.439c5.656-4.583 6.378-12.979 1.57-18.442C488.412 45.226 421.496 10.44 350.59.137c-7.243-1.057-13.925 4.118-14.662 11.393h-.016Z"/><path fill="#FF6900" d="M425.7 265.053c8.176 23.793 9.234 49.978 2.165 75.039-1.795 6.341 1.619 13.004 7.774 15.364l147.58 56.688c6.844 2.633 14.458-1.027 16.782-7.979 22.697-68.07 21.752-143.639-6.09-213.106-2.71-6.775-10.532-9.922-17.184-6.92L432.56 249.224c-6.01 2.713-9.008 9.568-6.876 15.813l.016.016Z"/><path fill="#3E4C60" d="m392.883 413.138 99.708 123.107c4.608 5.703 13.037 6.362 18.528 1.51a307.056 307.056 0 0 0 38.117-40.36c4.528-5.769 3.372-14.14-2.585-18.43l-128.593-92.53c-5.346-3.84-12.78-3.085-17.083 1.895a124.01 124.01 0 0 1-7.177 7.616c-4.72 4.595-5.058 12.067-.915 17.192Z"/></g></svg>
