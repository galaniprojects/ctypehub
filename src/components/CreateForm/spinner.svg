<svg xmlns="http://www.w3.org/2000/svg" height="74" width="74" viewBox="0 0 74 74">
  <path fill-opacity="0" stroke="#E7E7E7" stroke-width="3" d="M2,37 A35,35 0 1 0 37,2" />
  <path fill-opacity="0" stroke="#767676" stroke-width="3" d="M37,2 A35,35 0 0 0 2,37" />
</svg>
