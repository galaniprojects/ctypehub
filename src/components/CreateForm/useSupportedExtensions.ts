import { useEffect, useState } from 'react';

export function useSupportedExtensions() {
  const [extensions, setExtensions] = useState<
    Array<{ key: string; name: string }>
  >([]);

  useEffect(() => {
    function update() {
      const current = [...Object.entries(window.kilt)]
        .filter(([, api]) => 'signExtrinsicWithDid' in api)
        .map(([key, { name }]) => ({ key, name }));
      setExtensions(current);
    }

    update();
    window.dispatchEvent(new CustomEvent('kilt-dapp#initialized'));

    window.addEventListener('kilt-extension#initialized', update);
    return () => {
      window.removeEventListener('kilt-extension#initialized', update);
    };
  }, []);

  return extensions;
}
