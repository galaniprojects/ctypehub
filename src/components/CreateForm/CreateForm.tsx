import {
  FocusEvent,
  FormEvent,
  KeyboardEvent,
  MouseEvent,
  useCallback,
  useState,
} from 'react';
import { web3FromSource } from '@polkadot/extension-dapp';
import {
  Blockchain,
  connect,
  CType,
  Did,
  disconnect,
  KiltAddress,
} from '@kiltprotocol/sdk-js';

import styles from './CreateForm.module.css';

import { InjectedAccount, SelectAccount } from '../SelectAccount/SelectAccount';
import { Modal } from '../Modal/Modal';
import { PropertyFields } from '../PropertyFields/PropertyFields';
import { getProperties } from '../PropertyFields/getProperties';
import { offsets } from '../../utilities/offsets';
import { getBlockchainEndpoint } from '../../utilities/getBlockchainEndpoint';
import { generatePath, paths } from '../../paths';

import { useSupportedExtensions } from './useSupportedExtensions';

export function CreateForm() {
  const [propertiesCount, setPropertiesCount] = useState(0);
  const handleAddPropertyClick = useCallback(
    () => setPropertiesCount(propertiesCount + 1),
    [propertiesCount],
  );

  const [tags, setTags] = useState<string[]>([]);

  const addTag = useCallback(
    (tag: string) => {
      const trimmed = tag.trim().replace(/,/g, '').toLowerCase();
      if (!trimmed || trimmed.length < 2 || tags.includes(trimmed)) {
        return;
      }
      setTags([...tags, trimmed]);
    },
    [tags],
  );

  const handleTagInputKeyUp = useCallback(
    (event: KeyboardEvent<HTMLInputElement>) => {
      const { key, currentTarget } = event;

      if (![',', 'Backspace'].includes(key)) {
        return;
      }

      if (key === ',' && currentTarget.value) {
        addTag(currentTarget.value);
        currentTarget.value = '';
      }
      if (key === 'Backspace' && !currentTarget.value && tags.length > 0) {
        const lastTag = tags.slice(-1)[0];
        setTags(tags.slice(0, -1));
        currentTarget.value = lastTag;
      }
    },
    [addTag, tags],
  );

  const handleTagInputBlur = useCallback(
    (event: FocusEvent<HTMLInputElement>) => {
      addTag(event.currentTarget.value);
      event.currentTarget.value = '';
    },
    [addTag],
  );

  const handleRemoveTag = useCallback(
    (event: MouseEvent<HTMLButtonElement>) => {
      event.preventDefault();
      setTags(tags.filter((tag) => tag !== event.currentTarget.value));
    },
    [tags],
  );

  const [account, setAccount] = useState<InjectedAccount>();
  const [progress, setProgress] = useState(false);
  const [error, setError] = useState(false);
  const unsetError = useCallback(() => setError(false), []);

  const handleSubmit = useCallback(
    async (event: FormEvent<HTMLFormElement>) => {
      try {
        event.preventDefault();
        if (!account) {
          throw new Error('Cannot trigger submit without an account');
        }

        setProgress(true);
        const formData = new FormData(event.currentTarget);
        const title = formData.get('title') as string;
        const description = formData.get('description') as string | undefined;
        const values = [...formData.entries()] as [string, string][];

        const properties = getProperties(propertiesCount, values);
        const cType = CType.fromProperties(title, properties);
        const cTypeUrl = generatePath(paths.ctypeDetails, cType.$id);

        const api = await connect(getBlockchainEndpoint());
        const existing = await api.query.ctype.ctypes(
          CType.idToHash(cType.$id),
        );
        if (existing.isSome) {
          const redirect = window.confirm(
            'An identical CType already exists. Open its page?',
          );
          if (redirect) {
            window.location.href = cTypeUrl;
          }
          return;
        }

        const nativeEvent = event.nativeEvent as SubmitEvent;
        const submitter = nativeEvent.submitter as HTMLButtonElement;
        const extensionKey = submitter?.value;
        const extension = window.kilt[extensionKey];

        const createTx = api.tx.ctype.add(CType.toChain(cType));
        const authorized = await extension.signExtrinsicWithDid(
          createTx.toHex(),
          account.address as KiltAddress,
        );
        const creator = Did.parse(authorized.didKeyUri).did;
        const authorizedTx = api.tx(authorized.signed);

        const injected = await web3FromSource(account.meta.source);
        const signed = await authorizedTx.signAsync(account.address, injected);
        await Blockchain.submitSignedTx(signed);

        const extrinsicHash = signed.hash.toHex();
        const response = await fetch(paths.ctypes, {
          method: 'POST',
          body: JSON.stringify({
            cType,
            extrinsicHash,
            creator,
            description,
            tags,
          }),
          headers: { 'Content-Type': 'application/json' },
        });
        if (response.ok) {
          window.location.href = `${cTypeUrl}?success`;
          return;
        }

        console.error(response);
        setError(true);
      } finally {
        setProgress(false);
        await disconnect();
      }
    },
    [account, propertiesCount, tags],
  );

  const extensions = useSupportedExtensions();

  return (
    <section className={styles.container}>
      <form onSubmit={handleSubmit} className={styles.form}>
        <h3 className={styles.heading}>Create a Claim Type</h3>

        <p>
          To create a Claim Type you need a{' '}
          <a
            className={styles.anchor}
            href="https://kilt-protocol.org/get-did/index.html"
          >
            KILT DID
          </a>{' '}
          and the{' '}
          <a className={styles.anchor} href="https://www.sporran.org/">
            Sporran wallet
          </a>{' '}
          (or another wallet supporting the{' '}
          <a
            className={styles.anchor}
            href="https://github.com/KILTprotocol/spec-ext-didsign-api"
          >
            DID Sign API Specification v1.1+
          </a>
          )
        </p>

        <p>
          <label className={styles.label}>
            Title:
            <input className={styles.input} name="title" required />
          </label>
        </p>

        <p>
          <label className={styles.label}>
            Description (optional):
            <textarea className={styles.description} name="description" />
          </label>
        </p>

        {offsets(propertiesCount).map((index) => (
          <fieldset key={index} className={styles.fieldset}>
            <legend>Property {index + 1}</legend>
            <PropertyFields index={index} />
          </fieldset>
        ))}

        <fieldset className={styles.fieldset}>
          <legend>Property {propertiesCount + 1}</legend>
          <p>
            <button
              type="button"
              onClick={handleAddPropertyClick}
              className={styles.add}
            >
              Add Property ➕️
            </button>
          </p>
        </fieldset>

        <div className={styles.label}>
          <label htmlFor="tagInput">Tags (Optional):</label>
          <ul className={styles.tags}>
            {tags.map((tag) => (
              <li key={tag} className={styles.tag}>
                {tag}
                <button
                  type="button"
                  aria-label="remove tag"
                  className={styles.removeTag}
                  value={tag}
                  onClick={handleRemoveTag}
                />
              </li>
            ))}

            <li className={styles.tagInputContainer}>
              <input
                id="tagInput"
                className={styles.tagInput}
                onKeyUp={handleTagInputKeyUp}
                onBlur={handleTagInputBlur}
                maxLength={50}
                aria-describedby="tagInputDescription"
              />
            </li>
          </ul>
          <p id="tagInputDescription" className={styles.tagInputDescription}>
            Enter a comma after each tag
          </p>
        </div>

        <SelectAccount onSelect={setAccount} />

        {extensions.length === 0 && <p>No wallets detected</p>}
        {extensions.map(({ key, name }) => (
          <button type="submit" className={styles.submit} value={key} key={key}>
            Create Claim Type via {name}
          </button>
        ))}

        {progress && (
          <Modal>
            <p className={styles.progress}>
              <span className={styles.spinner} />
              Storing the CType on the KILT blockchain
            </p>
          </Modal>
        )}

        {error && (
          <Modal>
            <output className={styles.output}>Error: Transaction Failed</output>
            <button type="button" onClick={unsetError} className={styles.retry}>
              Try again
            </button>
          </Modal>
        )}
      </form>
    </section>
  );
}
