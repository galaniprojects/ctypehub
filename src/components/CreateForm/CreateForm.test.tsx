// @vitest-environment jsdom

import { describe, expect, it, vi } from 'vitest';
import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { CreateForm } from './CreateForm';
import { useSupportedExtensions } from './useSupportedExtensions';

vi.mock('./useSupportedExtensions');
vi.mocked(useSupportedExtensions).mockReturnValue([
  { key: 'sporran', name: 'Sporran' },
]);

describe('CreateForm', () => {
  it('should render', async () => {
    const { container, queryAllByLabelText, getByText, getByLabelText } =
      render(<CreateForm />);
    expect(container).toMatchSnapshot();

    await userEvent.click(getByText(/Add Property/));
    await waitFor(() => queryAllByLabelText('Type:').length === 1);
    expect(container).toMatchSnapshot('adding property');

    await userEvent.type(
      getByLabelText('Tags (Optional):'),
      'tag1  ,tag2, x, tag3,',
    );
    expect(container).toMatchSnapshot('with tags');
  });
});
