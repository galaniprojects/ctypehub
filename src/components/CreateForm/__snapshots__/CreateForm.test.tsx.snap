// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateForm > should render > adding property 1`] = `
<div>
  <section
    class="_container_eddc76"
  >
    <form
      class="_form_eddc76"
    >
      <h3
        class="_heading_eddc76"
      >
        Create a Claim Type
      </h3>
      <p>
        To create a Claim Type you need a
         
        <a
          class="_anchor_eddc76"
          href="https://kilt-protocol.org/get-did/index.html"
        >
          KILT DID
        </a>
         
        and the
         
        <a
          class="_anchor_eddc76"
          href="https://www.sporran.org/"
        >
          Sporran wallet
        </a>
         
        (or another wallet supporting the
         
        <a
          class="_anchor_eddc76"
          href="https://github.com/KILTprotocol/spec-ext-didsign-api"
        >
          DID Sign API Specification v1.1+
        </a>
        )
      </p>
      <p>
        <label
          class="_label_eddc76"
        >
          Title:
          <input
            class="_input_eddc76"
            name="title"
            required=""
          />
        </label>
      </p>
      <p>
        <label
          class="_label_eddc76"
        >
          Description (optional):
          <textarea
            class="_description_eddc76"
            name="description"
          />
        </label>
      </p>
      <fieldset
        class="_fieldset_eddc76"
      >
        <legend>
          Property 
          1
        </legend>
        <p>
          <label
            class="_label_8bc908"
          >
            Name:
            <input
              class="_input_8bc908"
              name="property[0].name"
              required=""
            />
          </label>
        </p>
        <p>
          <label
            class="_label_8bc908"
          >
            Type:
            <select
              class="_input_8bc908"
              name="property[0].type"
            >
              <option
                value="string"
              >
                string
              </option>
              <option
                value="integer"
              >
                integer
              </option>
              <option
                value="number"
              >
                number
              </option>
              <option
                value="boolean"
              >
                boolean
              </option>
              <option
                value="reference"
              >
                reference
              </option>
            </select>
          </label>
        </p>
        <fieldset
          class="_fieldset_8bc908"
        >
          <legend>
            Limitations (optional)
          </legend>
          <p
            class="_group_8bc908"
          >
            <label
              class="_array_8bc908"
            >
              <input
                name="property[0].limitations"
                type="radio"
                value="any"
              />
              any matching value
            </label>
            <label
              class="_array_8bc908"
            >
              <input
                name="property[0].limitations"
                type="radio"
                value="enum"
              />
              fixed set of values
            </label>
          </p>
        </fieldset>
        <p>
          <label
            class="_array_8bc908"
          >
            <input
              name="property[0].array"
              type="checkbox"
              value="array"
            />
            is an array of such values
          </label>
        </p>
      </fieldset>
      <fieldset
        class="_fieldset_eddc76"
      >
        <legend>
          Property 
          2
        </legend>
        <p>
          <button
            class="_add_eddc76"
            type="button"
          >
            Add Property
          </button>
        </p>
      </fieldset>
      <div
        class="_label_eddc76"
      >
        <label
          for="tagInput"
        >
          Tags (Optional):
        </label>
        <ul
          class="_tags_eddc76"
        >
          <li
            class="_tagInputContainer_eddc76"
          >
            <input
              aria-describedby="tagInputDescription"
              class="_tagInput_eddc76"
              id="tagInput"
              maxlength="50"
            />
          </li>
        </ul>
        <p
          class="_tagInputDescription_eddc76"
          id="tagInputDescription"
        >
          Enter a comma after each tag
        </p>
      </div>
      <p
        class="_label_1b3ac1"
      >
        <button
          class="_load_1b3ac1"
          type="button"
        >
          Detect accounts
        </button>
        and
         
        <label>
          select payment KILT account:
          <select
            class="_select_1b3ac1"
            required=""
          />
        </label>
      </p>
      <button
        class="_submit_eddc76"
        type="submit"
        value="sporran"
      >
        Create Claim Type via 
        Sporran
      </button>
    </form>
  </section>
</div>
`;

exports[`CreateForm > should render > with tags 1`] = `
<div>
  <section
    class="_container_eddc76"
  >
    <form
      class="_form_eddc76"
    >
      <h3
        class="_heading_eddc76"
      >
        Create a Claim Type
      </h3>
      <p>
        To create a Claim Type you need a
         
        <a
          class="_anchor_eddc76"
          href="https://kilt-protocol.org/get-did/index.html"
        >
          KILT DID
        </a>
         
        and the
         
        <a
          class="_anchor_eddc76"
          href="https://www.sporran.org/"
        >
          Sporran wallet
        </a>
         
        (or another wallet supporting the
         
        <a
          class="_anchor_eddc76"
          href="https://github.com/KILTprotocol/spec-ext-didsign-api"
        >
          DID Sign API Specification v1.1+
        </a>
        )
      </p>
      <p>
        <label
          class="_label_eddc76"
        >
          Title:
          <input
            class="_input_eddc76"
            name="title"
            required=""
          />
        </label>
      </p>
      <p>
        <label
          class="_label_eddc76"
        >
          Description (optional):
          <textarea
            class="_description_eddc76"
            name="description"
          />
        </label>
      </p>
      <fieldset
        class="_fieldset_eddc76"
      >
        <legend>
          Property 
          1
        </legend>
        <p>
          <label
            class="_label_8bc908"
          >
            Name:
            <input
              class="_input_8bc908"
              name="property[0].name"
              required=""
            />
          </label>
        </p>
        <p>
          <label
            class="_label_8bc908"
          >
            Type:
            <select
              class="_input_8bc908"
              name="property[0].type"
            >
              <option
                value="string"
              >
                string
              </option>
              <option
                value="integer"
              >
                integer
              </option>
              <option
                value="number"
              >
                number
              </option>
              <option
                value="boolean"
              >
                boolean
              </option>
              <option
                value="reference"
              >
                reference
              </option>
            </select>
          </label>
        </p>
        <fieldset
          class="_fieldset_8bc908"
        >
          <legend>
            Limitations (optional)
          </legend>
          <p
            class="_group_8bc908"
          >
            <label
              class="_array_8bc908"
            >
              <input
                name="property[0].limitations"
                type="radio"
                value="any"
              />
              any matching value
            </label>
            <label
              class="_array_8bc908"
            >
              <input
                name="property[0].limitations"
                type="radio"
                value="enum"
              />
              fixed set of values
            </label>
          </p>
        </fieldset>
        <p>
          <label
            class="_array_8bc908"
          >
            <input
              name="property[0].array"
              type="checkbox"
              value="array"
            />
            is an array of such values
          </label>
        </p>
      </fieldset>
      <fieldset
        class="_fieldset_eddc76"
      >
        <legend>
          Property 
          2
        </legend>
        <p>
          <button
            class="_add_eddc76"
            type="button"
          >
            Add Property
          </button>
        </p>
      </fieldset>
      <div
        class="_label_eddc76"
      >
        <label
          for="tagInput"
        >
          Tags (Optional):
        </label>
        <ul
          class="_tags_eddc76"
        >
          <li
            class="_tag_eddc76"
          >
            tag1  
            <button
              aria-label="remove tag"
              class="_removeTag_eddc76"
              type="button"
              value="tag1  "
            />
          </li>
          <li
            class="_tag_eddc76"
          >
            tag2
            <button
              aria-label="remove tag"
              class="_removeTag_eddc76"
              type="button"
              value="tag2"
            />
          </li>
          <li
            class="_tag_eddc76"
          >
            tag3
            <button
              aria-label="remove tag"
              class="_removeTag_eddc76"
              type="button"
              value="tag3"
            />
          </li>
          <li
            class="_tagInputContainer_eddc76"
          >
            <input
              aria-describedby="tagInputDescription"
              class="_tagInput_eddc76"
              id="tagInput"
              maxlength="50"
            />
          </li>
        </ul>
        <p
          class="_tagInputDescription_eddc76"
          id="tagInputDescription"
        >
          Enter a comma after each tag
        </p>
      </div>
      <p
        class="_label_1b3ac1"
      >
        <button
          class="_load_1b3ac1"
          type="button"
        >
          Detect accounts
        </button>
        and
         
        <label>
          select payment KILT account:
          <select
            class="_select_1b3ac1"
            required=""
          />
        </label>
      </p>
      <button
        class="_submit_eddc76"
        type="submit"
        value="sporran"
      >
        Create Claim Type via 
        Sporran
      </button>
    </form>
  </section>
</div>
`;

exports[`CreateForm > should render 1`] = `
<div>
  <section
    class="_container_eddc76"
  >
    <form
      class="_form_eddc76"
    >
      <h3
        class="_heading_eddc76"
      >
        Create a Claim Type
      </h3>
      <p>
        To create a Claim Type you need a
         
        <a
          class="_anchor_eddc76"
          href="https://kilt-protocol.org/get-did/index.html"
        >
          KILT DID
        </a>
         
        and the
         
        <a
          class="_anchor_eddc76"
          href="https://www.sporran.org/"
        >
          Sporran wallet
        </a>
         
        (or another wallet supporting the
         
        <a
          class="_anchor_eddc76"
          href="https://github.com/KILTprotocol/spec-ext-didsign-api"
        >
          DID Sign API Specification v1.1+
        </a>
        )
      </p>
      <p>
        <label
          class="_label_eddc76"
        >
          Title:
          <input
            class="_input_eddc76"
            name="title"
            required=""
          />
        </label>
      </p>
      <p>
        <label
          class="_label_eddc76"
        >
          Description (optional):
          <textarea
            class="_description_eddc76"
            name="description"
          />
        </label>
      </p>
      <fieldset
        class="_fieldset_eddc76"
      >
        <legend>
          Property 
          1
        </legend>
        <p>
          <button
            class="_add_eddc76"
            type="button"
          >
            Add Property
          </button>
        </p>
      </fieldset>
      <div
        class="_label_eddc76"
      >
        <label
          for="tagInput"
        >
          Tags (Optional):
        </label>
        <ul
          class="_tags_eddc76"
        >
          <li
            class="_tagInputContainer_eddc76"
          >
            <input
              aria-describedby="tagInputDescription"
              class="_tagInput_eddc76"
              id="tagInput"
              maxlength="50"
            />
          </li>
        </ul>
        <p
          class="_tagInputDescription_eddc76"
          id="tagInputDescription"
        >
          Enter a comma after each tag
        </p>
      </div>
      <p
        class="_label_1b3ac1"
      >
        <button
          class="_load_1b3ac1"
          type="button"
        >
          Detect accounts
        </button>
        and
         
        <label>
          select payment KILT account:
          <select
            class="_select_1b3ac1"
            required=""
          />
        </label>
      </p>
      <button
        class="_submit_eddc76"
        type="submit"
        value="sporran"
      >
        Create Claim Type via 
        Sporran
      </button>
    </form>
  </section>
</div>
`;
